triggers/daml/Daml/Trigger/LowLevel.daml

```diff
@@ -244,3 +255,111 @@ data Commands = Commands
   { commandId : CommandId
   , commands : [Command]
   }
+
+-- Same as in Script; TODO factor
+data Free f a
+  = Pure a
+  | Free (f (Free f a))
+
+instance Functor f => Functor (Free f) where
+  fmap f (Pure a) = Pure (f a)
+  fmap f (Free x) = Free (fmap f <$> x)
+
+instance Functor f => Applicative (Free f) where
+  pure = Pure
+  Pure f <*> Pure a = Pure (f a)
+  Pure f <*> Free x = Free (fmap f <$> x)
+  Free x <*> my     = Free ((<*> my) <$> x)
+
+instance Functor f => Action (Free f) where
+  Pure a >>= f = f a
+  Free x >>= f = Free ((>>= f) <$> x)
+
+liftFC : Functor f => f a -> Free f a
+liftFC fa = Free (Pure <$> fa)
+
+lowerFC : Action f => Free f a -> f a
+lowerFC (Pure a) = pure a
+lowerFC (Free x) = x >>= lowerFC
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
+  deriving Functor
+
+liftStateT : Functor m => m a -> StateT s m a
+liftStateT ma = StateT $ \s -> (,s) <$> ma
+
+instance Action m => Applicative (StateT s m) where
+  pure a = StateT (\s -> pure (a, s))
+  f <*> x = f >>= (<$> x)
+  
+instance Action m => Action (StateT s m) where
+  StateT x >>= f = StateT $ \s -> do
+    (x', s') <- x s
+    runStateT (f x') s'
+
+class ActionState s m | m -> s where
+  get : m s
+  put : s -> m ()
+
+modify : (Action m, ActionState s m) => (s -> s) -> m ()
+modify f = put . f =<< get
+
+instance Applicative m => ActionState s (StateT s m) where
+  get = StateT $ \s -> pure (s, s)
+  put s = StateT $ const $ pure ((), s)
+
+execStateT : Functor m => StateT s m a -> s -> m s
+execStateT (StateT fa) = fmap snd . fa
+
+zoom : Functor m => (t -> s) -> (t -> s -> t) -> StateT s m a -> StateT t m a
+zoom r w (StateT smas) = StateT $ \t ->
+  smas (r t) <&> \(a, s) -> (a, w t s)
+
+-- Must be kept in sync with Runner#handleStepFreeResult
+data TriggerF a =
+  GetTime (Time -> a)
+  | Submit (Commands, () -> a)
+  deriving Functor
+
+newtype TriggerSetup s a = TriggerSetup { runTriggerSetup : Free TriggerF a }
+  deriving (Functor, Applicative, Action)
+
+newtype TriggerRule s a = TriggerRule { runTriggerRule : StateT s (Free TriggerF) a }
+  deriving (Functor, Applicative, Action)
+
+deriving instance ActionState s (TriggerRule s)
+
+class HasTime m => ActionTrigger m where
+  liftTF : TriggerF a -> m a
+
+instance ActionTrigger (TriggerSetup s) where
+  liftTF = TriggerSetup . liftFC
+
+instance ActionTrigger (TriggerRule s) where
+  liftTF = TriggerRule . liftStateT . liftFC
+
+instance HasTime (TriggerSetup s) where
+  getTime = liftTF (GetTime identity)
+
+instance HasTime (TriggerRule s) where
+  getTime = liftTF (GetTime identity)
+
+submitCommands : ActionTrigger m => Commands -> m ()
+submitCommands cmds = liftTF (Submit (cmds, identity))
+
+type TriggerInitialState s = Party -> Time -> ActiveContracts -> (s, [Commands])
+type TriggerUpdate s = Time -> Message -> s -> (s, [Commands])
+
+forwardPortInitialState : TriggerInitialState s -> Party -> ActiveContracts -> Free TriggerF s
```

Comment:
Not that worried about trying to make it easy to migrate the low-level trigger API. Triggers are still alpha in general and the low-level API is even more likely to change.

Reply:
This was more for my own sanity-checking (that all ports must be possible) and my own ease of porting the integration tests. If it seems nicer to directly port the tests I'll get rid of them.